monkey/infection_monkey/telemetry/attack/victim_host_telem.py

```diff
@@ -0,0 +1,21 @@
+from infection_monkey.telemetry.attack.attack_telem import AttackTelem
+
+__author__ = "VakarisZ"
+
+
+class VictimHostTelem(AttackTelem):
+
+    def __init__(self, technique, status, machine):
+        """
+        ATT&CK telemetry that parses and sends VictimHost's (remote machine's) data
+        :param technique: Technique ID. E.g. T111
+        :param status: ScanStatus of technique
+        :param machine: VictimHost obj from model/host.py
+        """
+        super(VictimHostTelem, self).__init__(technique, status)
+        self.machine = {'domain_name': machine.domain_name, 'ip_addr': machine.ip_addr}
+
+    def get_data(self):
```

Comment:
Add a UT for this?

Reply:
This smells like a case of [Horizontal Code Reuse](http://propelorm.org/blog/2011/03/03/don-t-copy-code-oh-and-inheritance-and-composition-are-bad-too.html) - are you sure this is how you want to do this?

Reply:
Haven't heard of pros/cons of this code pattern before. But even according to the link you sent, they advise against horizontal code reuse whenever the inheritance is done in an abuse way (e.g. Book inherits from Persistable). In our case, the inheritance makes much more sense. The classes represent telemetry types, so it makes sense that they inherit from other telemetry types and have the top base class be a BaseTelem.
I don't see any scenario where any of these classes will need to inherit from something else, or have this pattern hurt us in any other way.

Reply:
I agree with Itay, inheritance here is used to extend parent functionality, not to include some unrelated methods by abusing inheritance. What's the alternative? To use `AttackTelem` as a service here, but that would make a wacky syntax, no?

Reply:
I think I haven't explained myself well enough, so I'll try again. But before that I'll say that if @itaymmguardicore is confident that this pattern won't hurt us and it makes sense - then I don't want to waste time waxing philosophical design points on such a small part of the code. 

I have two main issues with this code. One is more about SRP and the other is about horizontal code reuse.  

The current inheritance structure is: `BaseTelem` ⬅ `AttackTelem` ⬅ `VictimHostTelem`.

About SRP (This is about the `BaseTelem` ⬅ `AttackTelem` part of the tree):
The responsibilities of BaseTelem is to be able to send itself (via get_data) to the monkey island. So in a sense, it really should be named SendableParsedData. Since it defines the TelemType property, it might be called SendableParsedTelemetry. Each inheritence here *extends* the class, therefore adding responsibilities. But the extension is only in the data parsing part. There are 2 responsibilities here: Sending and Telemetry Parsing. 
However, they are both pretty simple in implementation, so it might be OK to keep them together. 

About horizontal code reuse (This is about the `AttackTelem` ⬅ `VictimHostTelem` part of the tree): 
The lowest class (VictimHostTelem) "reuses" the code in AttackTelem. However generally it's not great to inherit for code reuse, it should be inheriting for polymorphism. Both AttackTelem and VictimHostTelem should be BaseTelems, and they should both use composition (in this case, in my opinion) to hold and parse the data. The actual parsing can be functions in  the module which they both import and use. No duplication, no horizontal code reuse, and much cleaner (and testable!) code.

Reply:
Overall I think it's good that you brought it up here, because the larger the code piece the harder it would be to discuss design approaches. I've googled a bit about horizontal code reuse and maybe I'm just dumb, but I can't figure out what problems exactly does it cause... Some articles write, that multiple classes might share the same functionality that can't be reused trough inheritance (For e.g. `Horse` and `Tractor` classes both implement `pull_plow()`, but can't inherit from  `PlowPuller` class because `Horse` already inherits from `Animal` and `Tractor` inherits from `Vechicle`). But in python we can inherit multiple classes so that problem is non-existent (We can remove `pull_plow()` duplication by creating `PlowPuller` class and then do `class Horse(Animal, PlowPuller)`). Also, I can't figure out how is that a problem in testing. How is extended parent method harder to test than extended imported method? Lastly, lets say we create a separate module that `AttackTelem` and `VictimHostTelem` use. So `AttackTelem` uses that module, and `VictimHostTelem` uses that module in the same way(a.k.a. code duplication) + something extra. What does it solve?

Reply:
I understand what Shay was talking about now.
Me and Vakaris saw `VictimHostTelem` as an attack telemetry with host info.
Where Shay says, `VictimHostTelem` just requires the telemetry to have host info (even though in implementation it also requires attack info because of the inheritance).
What Shay suggested is that we'll be able to compose telemetries from several telemetry classes.
`VictimHostTelem` won't inherit from `AttackTelem`.
A composition of `VictimHostTelem` and `AttackTelem` would provide us with what we currently have.

I agree this allows even more code reuse. However, for 3 reasons I think we shouldn't do it:
1. Composition of several telemetry classes may reduce code duplication, but (in my opinion) over-complicates things. I've ran a lot of designs in my head how to implement this, and in all cases there are flaws. The need to deal with collision of names on composition of telemetries with the same name for different fields, the need to identify the type of the telemetry, and a need for easy creation of instances while keeping it all readable is very difficult.
2. Because of name collision of properties, the right thing to do under composition is to put all properties of a composition part under a new dictionary in the parsed data. This makes it non-backward compatible.
3. We really need to merge this already :)